File: src/DataStructure/Sort/InsertSort/BinarySort.java
package DataStructure.Sort.InsertSort;

/**
 * Created by phoebegl on 2017/5/1.
 * 二分法插入排序:稳定
 * O(n*log2n)
 */
public class BinarySort {

    public static void sort(int[] list,int left,int right) {
        int low,middle,high;
        int temp;
        for(int i = left+1;i<right;i++) {
            temp = list[i];
            low = left;
            high = i-1;
            while(low <= high) {
                middle = (low + high) / 2;
                if (list[i] < list[middle])
                    high = middle - 1;
                else
                    low = middle + 1;
            }

            for (int j = i-1;j>=low;j--)
                list[j+1] = list[j];
            list[low] = temp;
        }
    }
}

File: sort/binary_sort.py
# 二分插入排序
# input:13 24 3 56 34 3 78 12 29 49 84 51 9 100

def binary_sort(data):
    for i in range(len(data)):
        left, right = 0, i-1
        temp = data[i]
        while left <= right:
            mid = (left+right)//2
            if temp < data[mid]:
                right = mid-1
            else:
                left = mid +1
        j = i-1
        while j >= left:
            data[j+1] = data[j]
            j -= 1
        if left != i:
            data[left] = temp
    return data

def solution():
    data = [13,24,3,56,34,3,78,12,29,49,84,51,9,100]

    for i in range(1,len(data)):
        data = binary_sort(data)
    print(data)
solution()